### PR TITLE
[doorbird] Add support for version 2 encryption scheme

### LINKED ISSUE #15827: [doorbird] Upcoming API Changes
@mhilbush ,  we received this email from Doorbird .   Thanks!

Hi Openhab,

Affective November 2023 there is a new “v.2” handling for encrypting/decrypting events.  Version 1 has been deprecated and will be removed in the future.  It can also be disabled by the user in the administration area of the app.

Integrations which are using Version 1 should update as soon as possible.

Please see the [updated API](https://www.doorbird.com/downloads/api_lan.pdf?rev=0.34) for additional details.

EVENT MONITORING (UDP BROADCASTS) Pages 28 – 31

### PR DESCRIPTION
**Pull request by mhilbush**
Resolves #15827

### CODE REVIEW

File: bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/api/DoorbirdAPI.java
Comment (lsiepel) [inline bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/api/DoorbirdAPI.java:111]:
```suggestion
            logger.warn("Unable to communicate with Doorbird: {}", e.getMessage());
```
Don't you also want to set the thing status to offline communication eror? If so, it can be set to debug, as it provides some more detail then the thing status

File: bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/api/DoorbirdAPI.java
Comment (lsiepel) [inline bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/api/DoorbirdAPI.java:113]:
```suggestion
            logger.warn("Unable to parse Doorbird response: {}", e.getMessage());
```
Don't you also want to set the thing status to offline communication eror? If so, it can be set to debug, as it provides some more detail then the thing status

File: bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/api/DoorbirdSession.java
Comment (lsiepel) [inline bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/api/DoorbirdSession.java:34]: Would it be possible to prevent this null suppression?

File: bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/listener/DoorbirdEvent.java
Comment (lsiepel) [inline bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/listener/DoorbirdEvent.java:284]:
```suggestion
            logger.debug("Decryption FAILED");
```

**Review by lsiepel — CHANGES_REQUESTED** [review]
Left some minor comments, otherwise LGTM.

File: bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/api/DoorbirdSession.java
Comment (mhilbush) [reply bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/api/DoorbirdSession.java:34]: I'm not sure what to do about this. `fromJson` can return null, but Eclipse doesn't think so. So when i remove the null suppression, Eclipse complains about a redundant null check on line 37. I guess I could remove the null suppression, but it bothers me when Eclipse complains lol. I can go either way, so whatever you want to do is fine with me.

File: bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/listener/DoorbirdEvent.java
Comment (mhilbush) [reply bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/listener/DoorbirdEvent.java:284]: I left this as TRACE, because a failure here is "normal". The comment directly above explains why.

File: bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/api/DoorbirdAPI.java
Comment (mhilbush) [reply bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/api/DoorbirdAPI.java:111]: I'll change this and similar methods to to use `warn` instead of `info`.

As for setting the thing OFFLINE, I'd prefer to hold off on that as it requires changes in a number of other places to make sure it is set back ONLINE properly. And, since the device is on the users local network, the likelihood of these errors is pretty low.

WDYT?

File: bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/api/DoorbirdAPI.java
Comment (mhilbush) [reply bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/api/DoorbirdAPI.java:113]: See my comment above.

**Comment by mhilbush** [comment]
@lsiepel  Thanks for your feedback! Once I get your thoughts on my comments above, I can push the changes.

BTW, congrats on being named maintainer!

File: bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/api/DoorbirdSession.java
Comment (lsiepel) [reply bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/api/DoorbirdSession.java:34]: Null annotations can be a pain. :-/ As `fromJson` is not properly annotated, this is an acceptable warning. I prefer to have a warning over a suppresion, as that may lead to future issues.

File: bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/api/DoorbirdAPI.java
Comment (lsiepel) [reply bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/api/DoorbirdAPI.java:111]: A warn message may allready be a good step forward for users to troubleshoot connectivity issues.
That being said, i would recommend to have a proper thing status handling. I also understand that this PR's scope is adding a V2 API and not improving the allready existing thangstatus handling. So maybe we should leave that to a future PR.

File: bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/listener/DoorbirdEvent.java
Comment (lsiepel) [reply bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/listener/DoorbirdEvent.java:284]: Ok, as you carefully chosen this to be trace, it is fine by me.
I was somewhat triggered because earlier in this method there is a 
```
logger.debug("Decrypting v2 event, size of buffer: {}", bb.capacity());
``` 
and later on there is only a follow up in trace mode.

File: bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/api/DoorbirdAPI.java
Comment (mhilbush) [reply bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/api/DoorbirdAPI.java:111]: Ok, thanks.

File: bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/api/DoorbirdSession.java
Comment (mhilbush) [reply bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/api/DoorbirdSession.java:34]: Fair enough. I'll remove the null suppression.

File: bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/listener/DoorbirdEvent.java
Comment (mhilbush) [reply bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/listener/DoorbirdEvent.java:284]: Yes, I understand. Thanks. Given the wonky nature of how this works, it wasn't possible to have clear indication of an actual failure.

**Comment by mhilbush** [comment]
@lsiepel I pushed the changes. Thanks for your review.

**Review by lsiepel — APPROVED** [review]
Thanks! LGTM

**Comment by lsiepel** [comment]
As from the related issue i understand the V1 version is deprecated, but not yet taken down, so i guess there is no need for backporting to 4.1.x

**Comment by mhilbush** [comment]
Good question. You are correct, v1 only is deprecated at this point. However, I'd suggest we back port this if possible, since users would prefer to configure their Doorbird to send only the v2 encrypted packet. According to the vendor, sending both v1 and v2 has some performance impact on the relatively underpowered Doorbird device.

**Comment by mhilbush** [comment]
@lsiepel WDYT about cherry-picking this for 4.1.x?

**Comment by lsiepel** [comment]
>.......According to the vendor, sending both v1 and v2 has some performance impact on the relatively underpowered Doorbird device........

Done

**Comment by florian-h05** [comment]
Works fine with my D210FV, thank you very much!

**Pull request merged**
